File: bongf/java/src/ch17_shortest_path/Q40_HideAndSeek.java
package ch17_shortest_path;

import java.util.*;

public class Q40_HideAndSeek {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int m = sc.nextInt();
        // 동빈과 다른 점. 동빈은 Place를 넣어줬고 난 그냥 어차피 거리는 1이니까 Integer로 인덱스를 넣어줌
        List<List<Integer>> path = new ArrayList<>();
        for (int i = 0; i <= n; i++) {
            path.add(new ArrayList<>());
        }
        for (int i = 0; i < m; i++) {
            int x = sc.nextInt();
            int y = sc.nextInt();
            path.get(x).add(y);
            path.get(y).add(x);
        }

        int[] d = new int[n + 1];
        Arrays.fill(d, Integer.MAX_VALUE);
        d[0] = 0;
        d[1] = 0;

        PriorityQueue<Place> pq = new PriorityQueue<>();
        pq.add(new Place(1, 0));

        while (!pq.isEmpty()) {
            Place now = pq.poll();
            if (d[now.index] < now.distance) {
                continue;
            }
            for (Integer togo : path.get(now.index)) {
                int cost = now.distance + 1;
                if (cost < d[togo]) {
                    d[togo] = cost;
                    pq.add(new Place(togo, cost));
                }
            }
        }

        int maxDistance = 0;
        int placeToHide = 0;
        List<Integer> toHide = new ArrayList<>();
        for (int i = 1; i <= n; i++) {
            if (maxDistance < d[i]) {
                maxDistance = d[i];
                placeToHide = i;
                toHide = new ArrayList<>();
                toHide.add(i);
            } else if (maxDistance == d[i]) {
                toHide.add(i);
            }
        }
        System.out.print(placeToHide + " ");
        System.out.print(maxDistance + " ");
        System.out.print(toHide.size());
    }
}

class Place implements Comparable<Place> {
    int index;
    int distance;

    public Place(int index, int distance) {
        this.index = index;
        this.distance = distance;
    }

    @Override
    public int compareTo(Place other) {
        return this.distance - other.distance;
    }
}

/*
6 7
3 6
4 3
3 2
1 3
1 2
2 4
5 2
 */

File: bongf/python/ch17_shortest_path/q40_hide_and_seek.py
import heapq
def solution():
    INF = int(1e9)
    n, m = map(int, input().split(" "))
    path = [[] for _ in range(n+1)]
    for _ in range(m):
        x, y = map(int, input().split(" "))
        path[x].append((y,1))
        path[y].append((x,1))

    d = [INF] * (n+1)
    d[0] = 0

    q = []
    heapq.heappush(q, (0, 1))
    d[1] = 0
    while q:
        distance, now = heapq.heappop(q)
        if(d[now] < distance):
            continue
        for des in path[now]:
            cost = distance + des[1]
            if(cost < d[des[0]]):
                d[des[0]] = cost
                heapq.heappush(q, (cost, des[0]))
    
    '''
    내가 쓴 방법은 어찌되었든 for문을 2번 도는 형식인데 
    동빈이는 for문을 한 번 도는 것으로 해결 
    max_node = 0
    max_distance = 0
    result = []
    for i in range(1, n+1):
        if max_distance < d[i] :
            max_distance = d[i]
            max_node = i
            result = [max_node]
        elif max_distance == d[i]:
            result.append(i)
    print(max_node, max_distance, len(result))
    max_node를 바로 아웃해도 어차피 순서대로 for문을 도니까 제일 먼 노드 중에서 제일 작은 인덱스가 출력된다.
    '''
    answer_dis = max(d)
    to_hide = []
    for i in range(len(d)):
        if(d[i] == answer_dis) :
            to_hide.append(i)
    
    to_hide.sort()
    print(to_hide[0], end=" ")
    print(answer_dis, end=" ")
    print(len(to_hide))
        


solution()

'''
6 7
3 6
4 3
3 2
1 3
1 2
2 4
5 2
'''
